### PR TITLE
Long document entity extraction example

### PR DESCRIPTION
**Pull request by colin-openai**
Contains a notebook demonstrating how to iterate through a long document, extract entities from chunks and merge the results

### CODE REVIEW

File: examples/Entity_extraction_for_long_documents.ipynb [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+      "\n",
+      "0. Author: Tom Anderson (Page 1)\n",
+      "1.\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Example prompt - \n",
+    "document = '<document>'\n",
+    "template_prompt=f'''Extract key pieces of information from this regulation document. \n",
```
Comment (ted-at-openai) [inline examples/Entity_extraction_for_long_documents.ipynb:94]: can cut the trailing spaces on some of the lines in the templates

File: examples/Entity_extraction_for_long_documents.ipynb [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Consolidation\n",
+    "\n",
+    "We've been able to extract the first two answers safely, while the third was confounded by the date that appeared on every page, though the correct answer is in there as well.\n",
+    "\n",
+    "To tune this further you can consider experimenting with:\n",
+    "- A more descriptive or specific prompt\n",
+    "- Negative prompting, telling the model to ignore a particular thing i.e. the footnote date on every page\n",
```
Comment (ted-at-openai) [inline examples/Entity_extraction_for_long_documents.ipynb:336]: Not sure if negative prompting will work very well. Negative instructions are usually far less reliable than positive. I tried it, and the model still extracted a lot of 2022s.

**Review by ted-at-openai — APPROVED** [review]

File: examples/Entity_extraction_for_long_documents.ipynb [hunk 3]
```diff
@@ -0,0 +1,370 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Long Document Content Extraction\n",
+    "\n",
+    "We often need GPT-3 to help us extract key figures, dates or other bits of important content from documents that are too big to fit into the context window. One approach for solving this is to chunk the document up and process each chunk separately, before combining into one list of answers. \n",
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/Entity_extraction_for_long_documents.ipynb]: Updated wording

File: examples/Entity_extraction_for_long_documents.ipynb [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+    "\n",
+    "In this notebook we'll run through this approach:\n",
+    "- We'll load in a long PDF and pull the text out\n",
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/Entity_extraction_for_long_documents.ipynb]: Rewritten according to this approach

File: examples/Entity_extraction_for_long_documents.ipynb [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+    "- This simple approach will then be extended to three more difficult questions"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Setup\n",
+    "\n",
+    "We'll take our PDF, a Formula 1 Financial Regulation document on Power Units, and extract the text from it for entity extraction. We'll use this to try to extract answers that are buried in the content."
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/Entity_extraction_for_long_documents.ipynb]: Restructured with text all in the introduction

File: examples/Entity_extraction_for_long_documents.ipynb [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Extract key pieces of information from this regulation document. \n",
+      "If a particular piece of information is not present, output \"Not specified\". \n",
+      "When you extract a key piece of information, include the closest page number.\n",
+      "Use the following format:\n",
+      "0. Author\n",
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/Entity_extraction_for_long_documents.ipynb:79]: Made all of them questions

File: examples/Entity_extraction_for_long_documents.ipynb [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+      "1.\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Example prompt - \n",
+    "template_prompt=f'''Extract key pieces of information from this regulation document. \n",
+    "If a particular piece of information is not present, output \\\"Not specified\\\". \n",
+    "When you extract a key piece of information, include the closest page number.\n",
+    "Use the following format:\\n0. Author\\n1. How is a minor overspend breach calculated\\n2. How is a major overspend breach calculated\\n3.Which years do these financial regulations apply to\\n\\nDocument: \\\"\\\"\\\"{document}\\\"\\\"\\\"\\n\\n0. Author: Tom Anderson (Page 1)\\n1.'''\n",
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/Entity_extraction_for_long_documents.ipynb]: Updated

File: examples/Entity_extraction_for_long_documents.ipynb (hunk 1 above)
Comment (colin-openai) [reply examples/Entity_extraction_for_long_documents.ipynb]: Updated

File: examples/Entity_extraction_for_long_documents.ipynb (hunk 2 above)
Comment (colin-openai) [reply examples/Entity_extraction_for_long_documents.ipynb]: Removed negative prompting as a recommendation

File: examples/Entity_extraction_for_long_documents.ipynb [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+      "2. What is the value of External Manufacturing Costs in USD: Not specified\n",
+      "3. What is the Capital Expenditure Limit in USD: Not specified\n"
+     ]
+    }
+   ],
+   "source": [
+    "chunks = []\n",
+    "results = []\n",
+    "\n",
+    "# chunk up the contract into consecutive 10k character chunks\n",
```
[the remark this answers is not in the document]
Comment (colin-openai) [reply examples/Entity_extraction_for_long_documents.ipynb:194]: I've changed it to use tokens

**Pull request merged**